### PR TITLE
Support Freebsd

### PR DESCRIPTION
**Pull request by mldulaney**
This patch introduces support for x86 on FreeBSD.  I note that I added the check for FreeBSD on aarch64 to fail in case this gets built on an aarch64 FreeBSD system.

### CODE REVIEW

**Review by Mizux — APPROVED** [review]

**Comment by vishwin** [comment]
Those sysctl names are probably wrong. On a Broadwell system, only `hw.instruction_sse` is present, yet the hardware actually supports much more. `cpuinfo` from `linprocfs` is correct, however, but FreeBSD binaries should not be reading from any Linuxulator interfaces.

**Comment by mldulaney** [comment]
Hi, any chance this could get merged?

File: src/cpuinfo_x86.c [hunk 1]
```diff
@@ -1244,6 +1246,13 @@ static void DetectSseViaOs(X86Features* features) {
   features->ssse3 = GetDarwinSysCtlByName("hw.optional.supplementalsse3");
   features->sse4_1 = GetDarwinSysCtlByName("hw.optional.sse4_1");
   features->sse4_2 = GetDarwinSysCtlByName("hw.optional.sse4_2");
+#elif defined(CPU_FEATURES_OS_FREEBSD)
+  features->sse = sysctlbyname("hw.instruction_sse", NULL, NULL, NULL, 1);
```
Comment (gchatelet) [inline src/cpuinfo_x86.c:1250]: As far as I understand, this is not how `sysctlbyname` is supposed to be used.
https://www.freebsd.org/cgi/man.cgi?query=sysctlbyname&manpath=FreeBSD+13.0-RELEASE+and+Ports

One need to set proper arguments.
Also returned value is `0` upon success and `-1` upon failures.
I believe we should rename `GetDarwinSysCtlByName` above and use it here.

File: src/cpuinfo_aarch64.c [hunk 2]
```diff
@@ -80,6 +80,11 @@
 #define DEFINE_TABLE_FEATURE_TYPE Aarch64Features
 #include "define_tables.h"
 
+#if defined(CPU_FEATURES_OS_FREEBSD)
+#error "FreeBSD not yet supported on this arch"
+#endif  // CPU_FEATURES_OS
```
Comment (gchatelet) [inline src/cpuinfo_aarch64.c:85]: `// CPU_FEATURES_OS_FREEBSD`

File: src/cpuinfo_x86.c [hunk 3]
```diff
@@ -107,6 +107,8 @@
 #error "Darwin needs support for sysctlbyname"
 #endif
 #include <sys/sysctl.h>
+#elif defined(CPU_FEATURES_OS_FREEBSD)
```
Comment (gchatelet) [inline src/cpuinfo_x86.c:110]: This should probably be merged with the Darwin logic above.

**Review by gchatelet — CHANGES_REQUESTED** [review]

**Comment by gchatelet** [comment]
I'm closing this PR, FreeBSD support will be handled by #163 which will be submitted as soon as we have FreeBSD support in our CI.

Thank you @mldulaney for the patch: it didn't make it to the repo but it got me interested in solving the issue so it's a worthwhile contribution! (also thank you @vishwin for the comment)

**Pull request closed without merging**